pkg/kubelet/kubelet.go

```diff
@@ -3070,3 +3052,88 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
 func (kl *Kubelet) UnprepareDynamicResources(ctx context.Context, pod *v1.Pod) error {
 	return kl.containerManager.UnprepareDynamicResources(ctx, pod)
 }
+
+// staticPodRegistration ensures that all static pods are registered to the apiserver
+// before node became ready.
+func (kl *Kubelet) initialStaticPodsRegistration() error {
+	// kubelet running in standalone mode does not register static pods to the apiserver.
+	if kl.kubeClient == nil {
+		return nil
+	}
+	// Check if initial static pod registration has already been completed.
+	if kl.initialStaticPodsRegistered.Load() {
+		return nil
+	}
+	if _, err := kl.GetNode(); apierrors.IsNotFound(err) {
+		return errors.New("unable to check static pods are registered because node is not registered yet")
+	}
+	staticPodToMirrorPodMap := kl.podManager.GetStaticPodToMirrorPodMap()
+	for staticPod, mirrorPod := range staticPodToMirrorPodMap {
+		if mirrorPod == nil || mirrorPod.DeletionTimestamp != nil {
+			return fmt.Errorf("static pod %s is not registered", klog.KObj(staticPod))
+		}
+	}
+
+	kl.initialStaticPodsRegistered.Store(true)
+	return nil
+}
+
+// Ensure Mirror Pod for Static Pod exists and matches the current pod definition.
+// The function logs and ignores any errors.
+func (kl *Kubelet) ensureMirrorPodExists(staticPod, mirrorPod *v1.Pod) {
+	if !kubetypes.IsStaticPod(staticPod) {
+		return
+	}
+	deleted := false
+	if mirrorPod != nil {
+		if mirrorPod.DeletionTimestamp != nil || !kubepod.IsMirrorPodOf(mirrorPod, staticPod) {
```

Comment:
Yes, you commented above and I added a check for `DeletionTimestamp` - https://github.com/kubernetes/kubernetes/pull/126870#discussion_r1788128543

With respect to this function, if `DeletionTimestamp` is set, then we first delete it and then create it below. So it will be cleared out.